File: plugins/python/template/base/app/middleware/error_middleware.py
import logging
import traceback
from collections.abc import Awaitable, Callable

from fastapi import Request, Response
from fastapi.responses import JSONResponse
from sqlalchemy.exc import SQLAlchemyError
from starlette.middleware.base import BaseHTTPMiddleware

from app.core.exception import AppException

logger = logging.getLogger(__name__)


class ErrorHandlingMiddleware(BaseHTTPMiddleware):
    async def dispatch(
        self, request: Request, call_next: Callable[[Request], Awaitable[Response]]
    ) -> Response:
        try:
            response = await call_next(request)
            return response

        except AppException as exc:
            logger.warning(
                f"Application error: {exc.error_code}",
                extra={
                    "error_code": exc.error_code,
                    "path": request.url.path,
                    "method": request.method,
                    "details": exc.details,
                },
            )

            return JSONResponse(
                status_code=exc.status_code,
                content=exc.detail,
            )

        except SQLAlchemyError as exc:
            logger.error(
                "Database error occurred",
                exc_info=True,
                extra={
                    "path": request.url.path,
                    "method": request.method,
                    "error_type": type(exc).__name__,
                },
            )

            return JSONResponse(
                status_code=500,
                content={
                    "error_code": "error.database.transaction-failed",
                    "message": "Database error occurred",
                    "details": {
                        "error_type": type(exc).__name__,
                        "error": str(exc),
                    },
                },
            )

        except Exception as exc:
            logger.error(
                "Unhandled exception occurred",
                exc_info=True,
                extra={
                    "path": request.url.path,
                    "method": request.method,
                    "error_type": type(exc).__name__,
                    "traceback": traceback.format_exc(),
                },
            )

            return JSONResponse(
                status_code=500,
                content={
                    "error_code": "error.internal.unexpected",
                    "message": "An unexpected error occurred",
                    "details": {
                        "error_type": type(exc).__name__,
                    },
                },
            )

File: plugins/node/template/base/src/middleware/error.middleware.ts

import { FastifyError, FastifyReply, FastifyRequest } from "fastify";
import { ZodError } from "zod";
import { env } from "../config/env";
import { AppException } from "../core/exception";
import { logger } from "../core/logger";

export function errorHandler(
  error: FastifyError | Error,
  request: FastifyRequest,
  reply: FastifyReply
) {
  if (error instanceof AppException) {
    logger.warn({ err: error.message, statusCode: error.statusCode }, `${request.method} ${request.url}`);
    return reply.status(error.statusCode).send(error.toJSON());
  }

  if (error instanceof ZodError) {
    return reply.status(400).send({
      error_code: "error.validation.invalid-input",
      message: "Validation failed",
      details: { fields: error.issues },
    });
  }

  const fastifyError = error as FastifyError;
  if (fastifyError.validation) {
    return reply.status(400).send({
      error_code: "error.validation.invalid-input",
      message: "Validation failed",
      details: { fields: fastifyError.validation },
    });
  }

  logger.error({ err: error.message, stack: error.stack }, `${request.method} ${request.url}`);

  return reply.status(fastifyError.statusCode || 500).send({
    error_code: "error.internal.unexpected",
    message: "An unexpected error occurred",
    ...(env.NODE_ENV === "development" && { details: { stack: error.stack } }),
  });
}
